File: pat1099.cpp
/*
	思路：通过中序遍历建树后通过层序遍历输出结果

*/


#include<iostream>
#include<vector>
#include<queue>
#include<algorithm>
using namespace std;

struct node{
	int value;
	int left;
	int right;
}Tree[110];

vector<int> origin;
int index = 0;
void inorder(int root) {
	if (root == -1) return;
	inorder(Tree[root].left);
	Tree[root].value = origin[index++];
	inorder(Tree[root].right);
}

queue<int> q;
void bfs(int root) {
	q.push(root);
	printf("%d", Tree[root].value);
	while (!q.empty()){
		int now = q.front();
		q.pop();
		int left = Tree[now].left;
		int right = Tree[now].right;
		if (left != -1) {
			q.push(left);
			printf(" %d", Tree[left].value);
		}
		if (right != -1) {
			q.push(right);
			printf(" %d", Tree[right].value);
		}
	}
}

int main() {

	int N;
	scanf("%d", &N);
	origin.resize(N);
	for (int i = 0; i < N; i++) {
		scanf("%d %d", &Tree[i].left, &Tree[i].right);
	}
	for (int i = 0; i < N; i++) {
		scanf("%d", &origin[i]);
	}

	sort(origin.begin(), origin.end());
	inorder(0);
	bfs(0);
	return 0;
}

File: pat1099.py
"""
    思路：通过中序遍历建立二叉树，再通过层序遍历输出
    注：数据结构如：[[1,6,58],...]表示为0号节点左孩子节点为1，右孩子节点为6，值为58

"""
N = int(input())
tree = []

for i in range(N):
    children = [int(x) for x in input().split()]
    tree.append(children)

origin = [int(x) for x in input().split()]
origin.sort()

index = 0
def inorder(root):
    global index
    if root == -1:
        return
    inorder(tree[root][0])
    tree[root].append(origin[index])
    index += 1
    inorder(tree[root][1])

queue = []
def bfs(root):
    queue.append(root)
    print(tree[root][2],end='')
    while(len(queue) != 0):
        now = queue.pop(0)
        left = tree[now][0]     #   左孩子节点
        right = tree[now][1]    #   右孩子节点   
        if left != -1:
            queue.append(left)
            print(" " + str(tree[left][2]), end='')
        if right != -1:
            queue.append(right)
            print(" " + str(tree[right][2]), end='')

inorder(0)
bfs(0)


